trie/triedb/pathdb/disklayer.go

```diff
@@ -0,0 +1,259 @@
+// returned if the node is not found.
+func (dl *diskLayer) Node(owner common.Hash, path []byte, hash common.Hash) ([]byte, error) {
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
+
+	if dl.stale {
+		return nil, errSnapshotStale
+	}
+	// Try to retrieve the trie node from the not-yet-written
+	// node buffer first. Note the buffer is lock free since
+	// it's impossible to mutate the buffer before tagging the
+	// layer as stale.
+	n, err := dl.buffer.node(owner, path, hash)
+	if err != nil {
+		return nil, err
+	}
+	if n != nil {
+		dirtyHitMeter.Mark(1)
+		dirtyReadMeter.Mark(int64(len(n.Blob)))
+		return n.Blob, nil
+	}
+	dirtyMissMeter.Mark(1)
+
+	// Try to retrieve the trie node from the clean memory cache
+	if dl.db.cleans != nil {
+		if blob := dl.db.cleans.Get(nil, hash.Bytes()); len(blob) > 0 {
+			cleanHitMeter.Mark(1)
+			cleanReadMeter.Mark(int64(len(blob)))
+			return blob, nil
+		}
+		cleanMissMeter.Mark(1)
+	}
+	// Try to retrieve the trie node from the disk.
+	var (
+		nBlob []byte
+		nHash common.Hash
+	)
+	if owner == (common.Hash{}) {
+		nBlob, nHash = rawdb.ReadAccountTrieNode(dl.db.diskdb, path)
+	} else {
+		nBlob, nHash = rawdb.ReadStorageTrieNode(dl.db.diskdb, owner, path)
+	}
+	if nHash != hash {
+		return nil, &UnexpectedNodeError{
+			typ:      "disk",
+			expected: hash,
+			hash:     nHash,
+			owner:    owner,
+			path:     path,
+		}
+	}
+	if dl.db.cleans != nil && len(nBlob) > 0 {
+		dl.db.cleans.Set(hash.Bytes(), nBlob)
+		cleanWriteMeter.Mark(int64(len(nBlob)))
+	}
+	return nBlob, nil
+}
+
+// update returns a new diff layer on top with the given state set.
+func (dl *diskLayer) update(stateRoot common.Hash, id uint64, blockNumber uint64, nodes map[common.Hash]map[string]*trienode.Node, states *triestate.Set) *diffLayer {
+	return newDiffLayer(dl, stateRoot, id, blockNumber, nodes, states)
+}
+
+// commit merges the given bottom-most diff layer into the node buffer
+// and returns a newly constructed disk layer. Note the current disk
+// layer must be tagged as stale first to prevent re-access.
+func (dl *diskLayer) commit(bottom *diffLayer, force bool) (*diskLayer, error) {
+	dl.lock.Lock()
+	defer dl.lock.Unlock()
+
+	// Construct and store the state history first. If crash happens
+	// after storing the state history but without flushing the
+	// corresponding states(journal), the stored state history will
+	// be truncated in the next restart.
+	if dl.db.freezer != nil {
+		err := writeStateHistory(dl.db.freezer, bottom, dl.db.config.StateLimit)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Mark the diskLayer as stale before applying any mutations on top.
+	dl.stale = true
+
+	// Store the root->id lookup afterwards. All stored lookups are
+	// identified by the **unique** state root. It's impossible that
+	// in the same chain blocks are not adjacent but have the same
+	// root.
+	if dl.id == 0 {
+		rawdb.WriteStateID(dl.db.diskdb, dl.rootHash, 0)
```

Comment:
The state with id 0(initial state) can be two cases:

- Zero state

Zero state is the default state when initialize the database. Genesis state actually corresponds to state 1, the first state transition from 0 to 1 is the changes made in genesis.

- Snap sync'd state

The initial state can also be the snap sync'd state. Once we finish the state sync, we just explicitly mark it as state 0 and then do state transition on top.

### Why do we need to explicitly store the root->mapping for state0?

The state is determined as recoverable or not by checking if state histories from id+1 to latest disk layer are available.
For example, if we want to revert the state to snap sync'd one, we must know the id of it is 0, then check histories in range of [1, disklayer.id] are all present.

Therefore, the root->mapping of state0 is necessary.

Reply:
Can I revert to before state 0? I.e. undo the snap sync commit and restart snap sync?

Reply:
You can, but with a different method `Reset(types.EmptyStateRoot)`.

This is a special case in `SetHead`, if the target is not available, we 
just rewind our chain to genesis and redo the state transitions on top
(can be snap sync or full sync).

However, since in path-based, the genesis state is not reserved(state
overwrite), the only solution we can have is to reset the entire pathdb,
commit the genesis state transition on top(state transition 1), and then
relaunch the sync process.

Reply:
This can also be achieved by Recover, only if we have the full history since
genesis. 

For all state transitions we made, is from a null state. In another word, for
genesis state, it's the first valid state transition, from null->genesis state,
with id as 1.

If we still have the history 1, then it's possible to directly revert the state
to null.

However, if we either (a) snap sync (b) history corresponds to genesis is
pruned, then it's unfeasible.